### PR TITLE
Decompile ST0 func_801AF774

### PR DESCRIPTION
**Pull request by bismurphy**
Big function, does a bunch of stuff with DRAWENV which is neat.

Had to create an inline for addition, otherwise all my `lh` instructions were becoming `lhu`. Maybe there's a better solution, I'm not sure.

This is going out to a new file due to weird rodata stuff, I'm hoping decompiling EntityDraculaFinalForm (which I have a working scratch of) will allow everything to settle into place, but we'll see.

### CODE REVIEW

**Review by sozud — APPROVED** [review]
LGTM, thinking @Xeeynamo may want to look at the `add` thing

File: src/st/st0/2F774.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            prim->priority = 0xA8;
+            prim = prim->next;
+            self->ext.et_801AF774.prim4 = prim;
+            if (g_api.func_800EDB08(prim) == NULL) {
+                g_api.FreePrimitives(primIndex);
+                self->step = 0;
+                self->flags &= ~FLAG_HAS_PRIMS;
+                return;
+            }
+            prim->type = 7;
```
Comment (Xeeynamo) [inline src/st/st0/2F774.c:67]:
```suggestion
            prim->type = PRIM_ENV;
```
This is a pretty rare primitive type

File: src/st/st0/2F774.c [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+            prim->x0 = prim->x2 = self->posX.i.hi - 0x40;
+            prim->x1 = prim->x3 = self->posX.i.hi + 0x40;
+            prim->y0 = prim->y1 = self->posY.i.hi - 0x25;
+            prim->y2 = prim->y3 = self->posY.i.hi + 0x5B;
+            prim->r0 = prim->g0 = prim->b0 = 0x80;
+            LOW(prim->r1) = LOW(prim->r0);
+            LOW(prim->r2) = LOW(prim->r0);
+            LOW(prim->r3) = LOW(prim->r0);
+            prim->drawMode = 0xA;
```
Comment (Xeeynamo) [inline src/st/st0/2F774.c:57]:
```suggestion
            prim->drawMode = DRAW_UNK02 | DRAW_HIDE;
```
let's start documenting `drawMode`

File: src/st/st0/2F774.c [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    s32 xBase;
+    s32 yBase;
+    inline s32 add(a, b) { return a + b; }
+
+    switch (self->step) {
+    case 0:
+        InitializeEntity(D_801805E0);
+        self->hitboxState = 0;
+        self->animCurFrame = 0;
+        self->drawMode |= 0x30;
```
Comment (Xeeynamo) [inline src/st/st0/2F774.c:24]:
```suggestion
        self->drawMode |= DRAW_TPAGE | 0x20;
```
unfortunately 0x20 and 0x40 are not bitfields. It's kinda weird. It is documented in `RenderPrimitives`

File: src/st/st0/2F774.c [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            prim->drawMode = DRAW_HIDE;
+            prim = prim->next;
+            g_api.func_800EDB08(prim);
+            if (prim == NULL) {
+                g_api.FreePrimitives(primIndex);
+                self->step = 0;
+                self->flags &= ~FLAG_HAS_PRIMS;
+                return;
+            }
+            prim->type = 7;
```
Comment (Xeeynamo) [inline src/st/st0/2F774.c:77]:
```suggestion
            prim->type = PRIM_ENV;
```

**Review by Xeeynamo — APPROVED** [review]
Judging from the weird stuff happening here I think this is related to the "photo" taken after defeating Dracula. Especially due to the `tpage` going after `0x100`, which represents a 16-bit bitmap, I think. 

The inline function is pretty genius!

**Pull request merged**